### DISCUSSION TITLE
What's the best way to notify many waiting tasks?

### DISCUSSION THREAD

**Discussion by arthurprs** [tokio-rs/tokio#3718] [Q&A]
Consider an use-case where a producer wants to notify consumers about a change, even if they were doing something else at the time of the notification.

At first `Notify` and `Notify::notify_waiters` seem like a fit, but it's not. Because `notify_waiters` doesn't make the permit available for waiters that are _not_ waiting at the time of the notification.

C: subscribe to S by cloning the `Notify` 
P: `notifier.notify_waiters()`
C: `notifier.notified().await` <<< waits even if the permit was provided above

One way to get the expected behavior is to use a broadcast channel with capacity of 1.

C: subscribe to S by calling `sender.subscribe()` 
P: `sender.send(())`
C: `receiver.recv().await` <<< wakes up correctly, with either `Err(Lagged)` or `Ok(())`

Is there a better way of achieving this behavior?

I did find  #3066 but it was closed with a partial(?) solution.

**Comment by Darksonn** [comment]
I usually recommend a watch channel.

**Reply by arthurprs** [reply]
A `Watch` works too :thinking:  It's just a bit unergonomic to keep both `Sender` and `Receivers` around, as there's no equivalent to `subscribe()`

**Reply by taiki-e** [reply]
> It's just a bit unergonomic to keep both `Sender` and `Receivers` around, as there's no equivalent to `subscribe()`

I think we can expose `Sender::subscribe` (it is currently implemented as a private method).

https://github.com/tokio-rs/tokio/blob/a08ce0d3e06d650361283dc87c8fe14b146df15d/tokio/src/sync/watch.rs#L413

**Reply by arthurprs** [reply]
I think that would create a close to ideal component for this pattern :+1: 

I still have the felling that `Notifier` should support such pattern. The current `notify_waiters` is very easy to misuse and waiters end up waiting forever.

**Reply by arthurprs** [reply]
Opened #3757 so this is more visible.
